article/app/topmentions/TopMentionsS3Client.scala

```diff
@@ -0,0 +1,99 @@
+package topmentions
+
+import com.amazonaws.services.s3.AmazonS3
+import com.amazonaws.services.s3.model.{GetObjectRequest, S3Object}
+import com.amazonaws.util.IOUtils
+import common.GuLogging
+import conf.Configuration
+import play.api.libs.json.{JsError, JsSuccess, Json}
+import services.S3
+import topmentions.S3ObjectImplicits.RichS3Object
+import topmentions.TopMentionsResponse._
+
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+
+trait TopMentionsS3Client {
+  def getListOfKeys(): Future[List[String]]
+  def getObject(key: String): Future[TopMentionsDetails]
+}
+
+final class TopMentionsS3ClientImpl extends TopMentionsS3Client with S3 with GuLogging {
+  lazy val optionalBucket: Option[String] = Configuration.aws.topMentionsStoreBucket
+
+  def getListOfKeys(): Future[List[String]] = {
+    getClient { client =>
+      Try {
+        val s3ObjectList = client.listObjects(getBucket).getObjectSummaries().asScala.toList
+        s3ObjectList.map(_.getKey)
+      } match {
+        case Success(value) =>
+          log.info(s"got list of ${value.length} top mentions from S3")
+          Future.successful(value)
+        case Failure(exception) =>
+          log.error(s"failed in getting the list of top mentions from S3", exception)
+          Future.failed(exception)
+      }
+    }
+  }
+
+  def getObject(key: String): Future[TopMentionsDetails] = {
+    getClient { client =>
+      Try {
+        val request = new GetObjectRequest(getBucket, key)
+        client.getObject(request).parseToTopMentionsDetails
+      }.flatten match {
+        case Success(value) =>
+          log.info(s"got topMentionResponse from S3 for key ${key}")
+          Future.successful(value)
+        case Failure(exception) =>
+          log.error(s"S3 retrieval failed for key ${key}", exception)
+          Future.failed(exception)
+      }
+    }
+  }
+
+  private def getBucket() = {
+    optionalBucket.getOrElse(
+      throw new RuntimeException("top mention bucket config is empty, make sure config parameter has value"),
```

Comment:
Might be worth seeing if this can be wrapped in a Try, although perhaps its fine as is since its not really possible to recover from this.